Separate out publishing to github

cailafinn opened the pull request:
**Description of work.**

Jenkins would report a failure if the publishing to github failed because a repo wasn't given. But if you don't want to publish to github that's a valid entry.

**To test:**
1. Use the `build_packages_from_branch` pipeline with the following settings:
```
BUILD_DEVL = none
BUILD_PACKAGE = linux64    // It's quickest
PACKAGE_SUFFIX = publishtestgh
PUBLISH_PACKAGES = true
ANACONDA_CHANNEL_LABEL = publishtestconda
GITHUB_RELEASES_REPO = 
GITHUB_RELEASES_TAG =
GH_ORGANIZATION_OR_USERNAME = ConorMFinn
BRANCH_NAME = jenkins_github_publish
```
2. Check that the jenkins job passed and successfully published to condaforge without publishing to github. 

*There is no associated issue.*



*This does not require release notes* because **this is related only to our build servers.**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Separate out publishing to github

Jenkins would report a failure if the publishing to github failed
because a repo wasn't given. But if you don't want to publish to github
that's a valid entry.

martyngigg commented on buildconfig/Jenkins/Conda/nightly_build_and_deploy.jenkinsfile:
```diff
@@ -227,6 +226,26 @@ pipeline {
+      when {
+        beforeAgent true
+        beforeOptions true
+        expression { env.PUBLISH_PACKAGES == 'true' }
```
Thanks for this. It's a useful update.

The default behaviour for multiple expressions is for all to need to be true (https://www.jenkins.io/doc/book/pipeline/syntax/#when) so I assume that's what was meant? Could we make clear here that it requires that with an `allOf` block?

Commit:
Make when statements explicitly allOf

Everything in a where clause if allOf by default, but better to make it
crystal clear.

martyngigg commented:
I've started a [test build](https://builds.mantidproject.org/job/build_packages_from_branch/218/)

martyngigg reviewed (approved):
All looks good. The anaconda publish worked and GitHub publishing was skipped.
